modules/README.md

```diff
@@ -1,58 +1,164 @@
-                    "properties": {
-                        "name": {
-                            "type": "string"
-                        },
-                        "version": {
-                            "type": "string"
-                        },
+        "type" : "object",
+        "description" : "The action argument passed via stdin",
+        "properties" : {
+            "input" : {
+                "type" : "object",
+                "description" : "PXP request arguments (complies with action's input in metadata)"
+            },
+            "configuration" : {
+                "type" : "object",
+                "description" : "Module's configuration (complies with metadata's configuration)"
+            },
+            "output_files" : {
+                "type" : "object",
+                "description" : "Contains the paths of files for storing the action output",
+                "properties" : {
+                    "stdout" : {
+                        "type" : "string",
+                        "description" : "File for storing the output on stdout"
                     },
-                    "required": [ "name" ]
-                },
-                "output": {
-                    "type": "object",
-                    "properties": {
-                        "success": {
-                            "type": "boolean"
-                        }
+                    "stderr" : {
+                        "type" : "string",
+                        "description" : "File for storing the output on stderr"
                     },
-                    "required": [ "success" ]
-                }
+                    "exitcode" : {
+                        "type" : "string",
+                        "description" : "File for storing the action's exitcode"
+                    },
+                },
+                "required" : ["out", "err", "exitcode"],
+                "additionalProperties" : false,
             }
-         ]
-     }
+        },
+        "required" : ["input"],
+        "additionalProperties" : false,
+    }
+
+The `input` and `configuration` entries of an input object comply with the
+respective schemas included in the `metadata`; pxp-agent enforces that. Note
+that the `configuration` schema is optional; in case it's not included in the
+`metadata`, the content of the configuration file will be included in the input
+object if its format is valid JSON.
 
+Each action should write its processing results on stdout. Error messages should
+go on stderr.
 
-Invocation of an action will supply parameters as a document on stdin
+If the input object parsed from stdin contains the `output_files` object, the
+action must ALSO write its results, error messages, and exit code in the
+specified files (respectively: `stdout`, `stderr`, and `exitcode`). In this
+case, pxp-agent guarantees that those files will be read only after the
+`exitcode` file is created; the existence of `exitcode` will be used to indicate
```

Comment:
Yep, that's formally correct, but that's not the only indication of the action execution that pxp-agent uses. Consider that:

- pxp-agent will normally get the outcome of the action run directly by the leatherman::execution interface, since it must validate the results output and, if requested, send a final non-blocking response;
- only in case pxp-agent stops executing during an action run it will later rely on the existence of the exitcode file to determine if the action has made available its outcome;
- besides the pid file check and the current approach based on named mutexes to prevent (bigger) race conditions between leatherman::execution and onMessage handlers, pxp-agent can easily prevent concurrency issues use the existence of the file to determine the action execution state, then parse / validate stdout, process stderr, and finally get the exitcode int.

I believe this is purely an implementation issue; the above requirements (write on files) and guarantees (pxp-agent will check exitcode's existence before reading the rest) are enough to enable a safe workflow. Anyway, we have PCP-208 in program to improve the whole mechanism; things will change and it's not worth mandating a fourth file now.